staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/resourceclaim.go

```diff
@@ -24,32 +24,66 @@ limitations under the License.
 package resourceclaim
 
 import (
+	"errors"
 	"fmt"
 
 	v1 "k8s.io/api/core/v1"
 	resourcev1alpha2 "k8s.io/api/resource/v1alpha2"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+var (
+	// ErrAPIUnsupported is wrapped by the actual errors returned by Name and
+	// indicates that none of the required fields are set.
+	ErrAPIUnsupported = errors.New("none of the supported fields are set")
+
+	// ErrClaimNotFound is wrapped by the actual errors returned by Name and
+	// indicates that the claim has not been created yet.
+	ErrClaimNotFound = errors.New("ResourceClaim not created yet")
+)
+
 // Name returns the name of the ResourceClaim object that gets referenced by or
-// created for the PodResourceClaim. The name is deterministic and therefore
-// this function does not need any additional information and it will never
-// fail.
+// created for the PodResourceClaim. Three different results are possible:
+//
+//   - An error is returned when some field is not set as expected (either the
+//     input is invalid or the API got extended and the library and the client
+//     using it need to be updated) or the claim hasn't been created yet.
 //
-// Either podClaim.ResourceClaimName or podClaim.Template must be non-nil, but not
-// both. This is enforced by API validation.
+//     The error includes pod and pod claim name and the unexpected field and
+//     is derived from one of the pre-defined errors in this package.
+//
+//   - A nil string pointer and no error when the ResourceClaim intentionally
+//     didn't get created and the PodResourceClaim can be ignored.
+//
+//   - A pointer to the name and no error when the ResourceClaim got created.
+//     In this case the boolean determines whether IsForPod must be called
+//     after retrieving the ResourceClaim and before using it.
 //
 // If podClaim.Template is not nil, the caller must check that the
 // ResourceClaim is indeed the one that was created for the Pod by calling
 // IsUsable.
-func Name(pod *v1.Pod, podClaim *v1.PodResourceClaim) string {
-	if podClaim.Source.ResourceClaimName != nil {
-		return *podClaim.Source.ResourceClaimName
+func Name(pod *v1.Pod, podClaim *v1.PodResourceClaim) (name *string, mustCheckOwner bool, err error) {
```

Comment:
I used the opportunity to tighten validation in clients (controller manager, kubelet) which need to look up claims. This revised implementation of `Name` implements the "Consumers of this type must treat an empty object as if it has an unknown value." part of the API description.

This was partly left to the caller before and not always done consistently.

Reply:
How much overhead is it to just always check for the owner? It just feels awkward to have this extra boolean returned from here. Or better yet, why can't we just check for the owner in here and return an error if the pod passed in is not the owner whenever it "must" be checked.

Reply:
Not all claims will have the pod as owner. Only those generated for the pod are required to have that.

The code returning the boolean doesn't have access to the claim object and therefore cannot check it. The usual pattern is "determine claim name, get claim, check that it is usable" - the latter then includes checking the ownership.

Reply:
OK, that makes sense with not having enough information to do the check from within the call, but `mustCheckOwner ` still feels out of place as a return value to a method called `Name()`.  Can we make it its own API call somehow?

Reply:
OK. I don't feel too strongly, so we can leave it as is.